File: libs/javascript/src/node.js
const Node = (value) => {
	return {
		value: value,
		children: [],
		path: new Set(),

		addVisited(node) {
			this.path.add(node);
		},

		addChild(childNode) {
			this.children.push(childNode);
		},

		setPath(nodePath) {
			if (nodePath.length <= this.path.size) this.path = nodePath;
		},

		removeChild(childNode) {
			this.children = this.children.filter((i) => i != childNode);
		}
	}
}

export default Node;

File: libs/golang/pkg/node.go
package pkg

type Node struct {
	Value    rune
	Location Point
}

type RunnerNode struct {
	Node
	Path     Path
	Children []RunnerNode
}

func (n *RunnerNode) AddChild(c RunnerNode) {
	n.Children = append(n.Children, c)
}

func (r *RunnerNode) SetPath(p Path) {
	r.Path = p
}

func NewRunnerNode(n Node) RunnerNode {
	rn := RunnerNode{
		Node: n,
		Path: make(Path),
	}
	return rn
}
